docs/rfcs/component-universal-telemetry.md

```diff
@@ -0,0 +1,212 @@
+are unified by the component implementation. For example, the `otlp` receiver is a singleton, so its telemetry is not specific to a signal.
+Similarly, the `memory_limiter` processor is a singleton, so its telemetry is not specific to a pipeline.
+
+## Auto-Instrumentation Mechanism
+
+The mechanism of telemetry capture should be _external_ to components. Specifically, we should observe telemetry at each point where a
+component passes data to another component, and, at each point where a component consumes data from another component. In terms of the
+component graph, every _edge_ in the graph will have two layers of instrumentation - one for the producing component and one for the
+consuming component. Importantly, each layer generates telemetry ascribed to a single component instance, so by having two layers per
+edge we can describe both sides of each handoff independently.
+
+Telemetry captured by this mechanism should be associated with an instrumentation scope corresponding to the package which implements
+the mechanism. Currently, that package is `service/internal/graph`, but this may change in the future. Notably, this telemetry is not
+ascribed to individual component packages, both because the instrumentation scope is intended to describe the origin of the telemetry,
+and because no mechanism is presently identified which would allow us to determine the characteristics of a component-specific scope.
+
+### Auto-Instrumented Metrics
+
+There are two straightforward measurements that can be made on any pdata:
+
+1. A count of "items" (spans, data points, or log records). These are low cost but broadly useful, so they should be enabled by default.
+2. A measure of size, based on [ProtoMarshaler.Sizer()](https://github.com/open-telemetry/opentelemetry-collector/blob/9907ba50df0d5853c34d2962cf21da42e15a560d/pdata/ptrace/pb.go#L11).
+  These are high cost to compute, so by default they should be disabled (and not calculated).
+
+The location of these measurements can be described in terms of whether the data is "consumed" or "produced", from the perspective of the
+component to which the telemetry is attributed. Metrics which contain the term "produced" describe data which is emitted from the component,
+while metrics which contain the term "consumed" describe data which is received by the component.
+
+For both metrics, an `outcome` attribute with possible values `success` and `failure` should be automatically recorded, corresponding to
+whether or not the corresponding function call returned an error. Specifically, consumed measurements will be recorded with `outcome` as
+`failure` when a call from the previous component the `ConsumeX` function returns an error, and `success` otherwise. Likewise, produced
+measurements will be recorded with `outcome` as `failure` when a call to the next consumer's `ConsumeX` function returns an error, and
+`success` otherwise.
+
+```yaml
+    otelcol.receiver.produced.items:
```

Comment:
Can we remove the "otelcol." prefix and use a meaningful scope instead?

Reply:
I don't think we can remove the prefix as the Otel semconv requires metric names to be defined in the global registry and do not conflict with each other

Reply:
> PS: I meant [scope](https://github.com/open-telemetry/opentelemetry-specification/blob/ca2a0d40865c400361e6d443d040e836c0a54b81/specification/glossary.md?plain=1#L166) not prefix.

Ah, my bad. Agree with Dmitrii that some prefix should stay. The scope name should also be set (but I think we can just follow the OTel Go convention and we don't need to cover it on this RFC)

Reply:
Having the `otelcol` prefix would make it easier for users transitioning from the current metrics to the new ones. 

Additionally, if there is a time where SDKs emit metrics about their telemetry pipelines, we could avoid accidentally aggregating metrics for collector pipelines with SDK pipelines (assuming they would be using the same metric names)

For these reasons i'm in favour of keeping the prefix and also including the scope. I guess if users feel strongly enough, views could always be configured to rename metrics

Reply:
I will update the RFC to state that signals should include a scope name as well.

Reply:
Question about scope name:

If the telemetry is captured by an internal package, should the scope name be that of the module? e.g. `go.opentelemetry.io/service` instead of `go.opentelemetry.io/service/internal/graph`

I ask because as an internal module, it seems likely that the package name could change even if the `service` module is stable, which would break consistency of telemetry between versions. I think pinning to the module name is only slightly less helpful but more stable.

Reply:
I agree that using the module name is preferable

Reply:
I've added a section calling out the requirement for an instrumentation scope.